pilot/pkg/model/service.go

```diff
@@ -481,8 +483,24 @@ type IstioEndpoint struct {
 	// Labels points to the workload or deployment labels.
 	Labels labels.Instance
 
-	// Address is the address of the endpoint, using envoy proto.
-	Address string
+	// Addresses is the address of the endpoint, using envoy proto:
+	// https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/endpoint/v3/endpoint_components.proto#config-endpoint-v3-endpoint-additionaladdress
+	// This change can support multiple addresses for an endpoint, then there are 2 use cases for it, such as
+	// 1. An endpoint can have both ipv4 or ipv6.
+	// 2. An endpoint can have multiple addresses, more then one IP addresses which are the same IP families.
+	// There should be some constraints below:
+	// 1. Each address of the endpoint must have the same metadata.
+	// 2. The IstioEndpoint key generated by its function Key() should be the unique identifier.
+	// 3. The IP address of field `address` in Envoy Endpoint is equal to the first address of this field.
+	// When the additional_addresses field is populated for EDS in Envoy configuration, there would be a Happy Eyeballs
```

Comment:
How does kubernetes behave? Seems likely not this?

Why, in Istio context, would we have both IPs but not have one reachable? Is this something we are expected to handle?

Reply:
Hi @howardjohn, what I know is that at least the `containerd` handles multiple IP families via use happy eyeballs.
```
Why, in Istio context, would we have both IPs but not have one reachable? Is this something we are expected to handle?
```
The original requirement for Istio Dual Stack is that: If a service in Istio and in K8s cluster is deployed as Dual Stack, then the service which is deployed as IPv4 only or IPv6 only in the same Istio should be reachable for the Dual Stack service mentioned before,  and the other single stack services can reach the Dual Stack service as well. 
So it is possible that there is an endpoint with Dual Stack and can only use one of the IP addresses to access the single stack service.